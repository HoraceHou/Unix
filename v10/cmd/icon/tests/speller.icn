procedure spell(n)
   local m
   n := integer(n)
   return (
      if n <= 12 then
         "0zero,1one,2two,3three,4four,5five,6six,7seven,8eight,_
            9nine,10ten,11eleven,12twelve," ? {
               tab(find(n))
               move(*n)
               tab(upto(","))
               }
      else if n <= 19 then
         spell(n[2] || "0") ?
            (if ="for" then "four" else tab(find("ty"))) || "teen"
      else if n <= 99 then
         "2twen,3thir,4for,5fif,6six,7seven,8eigh,9nine," ? {
            tab(upto(n[1]))
            move(1)
            tab(upto(",")) || "ty" ||
               if n[2] ~= 0 then "-" || spell(n[2])
            }
      else if n <= 999 then
         spell(n[1]) || " hundred" ||
            (if (m := n[2+:2]) ~= 0 then " and " || spell(m))
      else if n <= 999999999 then
         spell(n[1:-3]) ?
            (tab(find("thousand")) || (move(8) & "million") | "") ||
            tab(0) ||
            " thousand" ||
            (if (m := n[-3:0]) ~= 0 then ", " || spell(m))
      else
         "number too large"
      )
end

procedure spellw(n)
   write(n, "	", spell(n))
   return
end

procedure main()
   every spellw(1 to 25)
   every spellw(30 to 110 by 3)
   spellw(945123342)
   every spellw(10000000 to 10000500)
end
