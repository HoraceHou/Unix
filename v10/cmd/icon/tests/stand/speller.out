1	one
2	two
3	three
4	four
5	five
6	six
7	seven
8	eight
9	nine
10	ten
11	eleven
12	twelve
21	twenty-one
22	twenty-two
23	twenty-three
24	twenty-four
25	twenty-five
33	thirty-three
36	thirty-six
39	thirty-nine
42	forty-two
45	forty-five
48	forty-eight
51	fifty-one
54	fifty-four
57	fifty-seven
63	sixty-three
66	sixty-six
69	sixty-nine
72	seventy-two
75	seventy-five
78	seventy-eight
81	eighty-one
84	eighty-four
87	eighty-seven
93	ninety-three
96	ninety-six
99	ninety-nine
102	one hundred and two
105	one hundred and five
108	one hundred and eight
945123342	nine hundred and forty-five million, one hundred and twenty-three thousand, three hundred and forty-two
